boards: arm: add Arduino Giga

fabiobaltieri opened the pull request:
Add support for the Arduino Giga board, an STM32H747XI based development board in Arduino form factor, featuring external flash, SDRAM, Bluetooth and WiFi.

---

Note: CI failure is unrelated, fix out in https://github.com/zephyrproject-rtos/zephyr/pull/55341

fabiobaltieri committed:
boards: arm: add Arduino GIGA

Add support for the Arduino GIGA board, an STM32H747XI based development
board in Arduino form factor, featuring external flash, SDRAM, Bluetooth
and WiFi.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
CI hotfix: https://github.com/zephyrproject-rtos/zephyr/pull/55341 (pushed here as well, will rebase it out as it gets merged)

nordicjm commented on boards/arm/arduino_giga/doc/index.rst:
```diff
@@ -0,0 +1,156 @@
+| QSPI      | on-chip    | QSPI flash                          |
++-----------+------------+-------------------------------------+
+
+Other hardware features are not yet supported on Zephyr porting.
```
*port

fabiobaltieri replied:
Done.

nordicjm commented on boards/arm/arduino_giga/arduino_giga_m7.dts:
```diff
@@ -0,0 +1,240 @@
+
+	chosen {
+		zephyr,console = &usart1;
+		zephyr,shell-uart = &usart1;
```
`zephyr,uart-mcumgr` might be useful

fabiobaltieri replied:
Yup, added.

nordicjm commented on boards/arm/arduino_giga/arduino_giga_m7.dts:
```diff
@@ -0,0 +1,240 @@
+};
+
+&usart1 {
+	status = "okay";
+	pinctrl-0 = <&usart1_tx_pa9 &usart1_rx_pb7>;
+	pinctrl-names = "default";
+	current-speed = <115200>;
+};
+
+&usart6 {
+	status = "okay";
+	pinctrl-0 = <&usart6_tx_pg14 &usart6_rx_pc7>;
+	pinctrl-names = "default";
+	current-speed = <115200>;
+};
+
+&uart4 {
+	status = "okay";
+	pinctrl-0 = <&uart4_tx_ph13 &uart4_rx_pi9>;
+	pinctrl-names = "default";
+	current-speed = <115200>;
+};
+
+&uart7 {
+	status = "okay";
+	pinctrl-0 = <&uart7_tx_pf7 &uart7_rx_pa8
+		     &uart7_cts_pf9 &uart7_rts_pf8>;
+	pinctrl-names = "default";
+	current-speed = <115200>;
+	status = "okay";
+	hw-flow-control;
```
do 4 UARTs really need to be enabled out of the box?

nordicjm commented on boards/arm/arduino_giga/arduino_giga_m7.dts:
```diff
@@ -0,0 +1,240 @@
+			read-only;
+		};
+
+		code_partition: partition@40000 {
+			label = "code";
+			reg = <0x40000 0x000c0000>;
+		};
```
Follow existing mcuboot-style naming, call this `slot0_partition` with label as `image-0`

nordicjm commented on boards/arm/arduino_giga/arduino_giga_m7_defconfig:
```diff
@@ -0,0 +1,36 @@
+# Enable pin controller
+CONFIG_PINCTRL=y
+
+# Use zephyr,code-partition as flash offset
+CONFIG_USE_DT_CODE_PARTITION=y
```
Is this always required? If someone just wants to build hello_world without a bootloader, this should be placed at 0x40000?

fabiobaltieri replied:
Yeah the board comes with an Arduino bootloader that look for the application at 0x40000. I see how that is very confusing since it can also be used with the ROM bootloader which is going to not have the offset and wipe the factory one, and both support USB-DFU.

The intended use would be with the Arduino one since that's compatible with their IDE and all the other stuff, and that's what the `board.cmake` VID/PID points to, so this should stay, but I added a big note explaining the two bootloader story in the documentation page. Could not find a page on how to re-flash the factory bootloader but it should land there eventually.

nordicjm replied:
@tejlmand @mbolivar-nordic this presents an interesting conundrum with things like `west flash --recover` - what if a user has a bootloader (which is not theirs/outside the zephyr build system) and wants to flash their application to the board or set of applications and recover the device with exception to the bootloader? I.e. wants the storage partition cleared also to remove any saved state of previous application testing. Is this something we ever want to support?

fabiobaltieri replied:
Are you thinking about having a `--recover` option outside of the `nrfjprog` target? Like to pad the settings area to nulls? That would not be a bad idea, a use case I bumped into: the bluetooth HID  (and other ble samples) by default stores the bond information in flash and have no way of deleting the bond. If one unpairs the device then can't pair any other until the settings are cleared, either `nrfjprog -e`, or hacking the BLE shell in the sample or hack the code to call `ble_unpair`.

mbolivar-nordic replied:
@nordicjm yeah, it's a nice-to-have, but I have no plans to do anything like that

nordicjm commented on boards/arm/arduino_giga/doc/index.rst:
```diff
@@ -0,0 +1,156 @@
+.. _arduino_giga_board:
```
Image is too large, please use something like https://tinypng.com/

fabiobaltieri replied:
Donez.

fabiobaltieri commented:
Updates:
- added a note explaining that the board can use both the ST ROM USB-DFU bootloader and the Arduino USB-DFU bootloader and that this is setup for the Arduino one (did not realize that that was totally confusing until Jamie's comment, thanks!)
- renamed the whole thing to match the product name (added `_r1` in the target name, so they now look like `arduino_giga_r1_m7` and used the full product name in the text description, that is `Arduino GIGA R1 WiFi`)
- disabled a bunch of serial/i2c/spi extra ports, only left the common Arduino ones enabled by default
- shrank the board image

fabiobaltieri commented:
- added `zephyr,memory-region-mpu`
- still waiting for https://github.com/zephyrproject-rtos/zephyr/pull/55341

nordicjm commented on boards/arm/arduino_giga_r1/arduino_giga_r1_m4.dts:
```diff
@@ -0,0 +1,37 @@
+		zephyr,console = &usart2;
+		zephyr,shell-uart = &usart2;
+		zephyr,sram = &sram1;
+		zephyr,flash = &flash1;
```
Where does the code for this reside? i.e. `zephyr,code-partition =` ?

fabiobaltieri replied:
`flash1`, the MCU has two independent flash banks, in the default config the M7 core entry point is the start of `flash0` and the M4 one is `flash1`, I think that way they get unblocked access through the bus matrix, see:

![x](https://user-images.githubusercontent.com/891546/222754183-2739e26a-ac3a-4be5-8329-d9a002459c02.png)

It's really the same settings as `stm32h747i_disco`, good enough for running some samples but realistically if ones wants to use the second core properly there's more custom setup to do anyway.

fabiobaltieri commented:
Dropped the teensy CI fixup from this PR, taking longer than I thought to get that in, it'll fail CI for this one but I can always retry it once it's in, saves from having unrelated code here.

fabiobaltieri commented:
Hotfix in, this passes CI now.

de-nordic reviewed (approved):
tests/mgmt/mcumgr/fs_mgmt_hash are OK

carlescufi reviewed (approved):
LGTM, thanks!
